### DISCUSSION TITLE
SMTP clients can not submit mail when running on same host

### DISCUSSION THREAD

**Discussion by ontheair81** [docker-mailserver/docker-mailserver#4157] [Q&A] [answered]
I am using docker mailserver since years with multiple domains and accounts. Everything works great, just recently I noticed something where I need help:

I am running multiple servers with different applications which automatically send e-mails to me (notifications, status-updates, etc.). For example Nextcloud, Uptime-Kuma, Vaultwarden, and many more.
These applications are using docker mailserver, submitting their e-mails via SMTP.

This still works fine for applications which are running _on other_ hosts.
But applications running _on the same host as docker mailserver_ can not send e-mails anymore (which worked in the past).
In my case, Nextcloud and Vaultwarden are running on the same machine as docker mailserver. These applications can not connect to docker mailserver anymore to send mails.

Sadly, the logs of these applications are not helpful at all. In the logs of docker mailserver I can not see a connection attempt from these applications. Fail2ban is not blocking the ip of the server.

Any idea what I am missing? Any help is very appreciated!

**Comment by ontheair81** [comment]
Looks like this is a general problem with docker: The mailserver gets requests from the internet, but not from other containers on the same host.

See [this discussion](https://stackoverflow.com/questions/78185828/gitlab-docker-image-cant-reach-smtp-ports-of-my-mailserver) and the accepted answer.

He solved it by manually changing the firewall configuration on the docker host. Which could be really a mess and I don't have too much experience with firewalls.

Maybe there is a setting in docker, allowing the containers to connect to the mailserver on the same host?

**Reply by polarathene** [reply]
> Maybe there is a setting in docker, allowing the containers to connect to the mailserver on the same host?

You need to share more context regarding your other container deployment vs DMS.

How are you trying to route the connection between the two? It should absolutely work if you have the containers connected to the same subnet:

```yaml
networks:
  default:
    name: my-custom-network-name
    ipam:
      driver: default
      config:
        - subnet: "172.16.42.0/24"

services:
  dms:
    image: mailserver/docker-mailserver
    hostname: mail.example.com
    networks:
      default:
        aliases:
          - mail.alias.test
    # ...
```

```yaml
networks:
  default:
    name: my-custom-network-name
    external: true

services:
  nextcloud:
    # ...  
```

The above has one compose file define a network with a name that other compose files can reference as an external network. You can also create that network with `docker network create` command if you prefer, and it can automatically choose a subnet to assign from the subnet address pool for you if you don't want to specify one explicitly.

The `default` part overrides the default network each compose file creates, which separates their subnet from other compose project files. Without compose, all containers via `docker run` instead start on the legacy `docker0` bridge which differs from custom networks (`docker network create` / `compose.yaml`) a bit feature wise, you generally want the custom networks.

In addition you'll notice I configured an alias for the `dms` service associated to the default network. Like the service name and the `hostname`, your other containers on that same docker network can then connect through that (_Docker has it's own embedded DNS by default configured_). So nextcloud could be configured to send mail to `dms:587`, `mail.example.com:465`, `mail.alias.test:25`, you can use the ports with any of them, that's just an example.

What you may run into is if DNS were queried for `user@example.com`, and nextcloud looks up the DNS MX record for `example.com` while on a network with a DMS container that has `hostname: example.com`, it may hit that Docker DNS and think _"I know the IP for `example.com`, there's this A record for it, but no MX record"_, which I think it'll still use anyway as fallback 🤷‍♂️ but that might be an issue if DNS queries do play out that way. I know they do if I locally host CoreDNS for a domain that has public DNS available, instead of forwarding to public DNS when the record is not found, if it found the domain configured it just stops there.

That said, usually you don't want to send mail to port 25, you can configure the service to submit mail over port 587 or 465 with credentials, and it'll connect direct to DMS to send the mail to an email address it manages or send outbound if not.

**Comment by casperklein** [comment, marked as answer]
> I can not see a connection attempt from these applications. Fail2ban is not blocking the ip of the server.

What IP address are you talking about? The private docker one or the servers public IP address?

The fact that you can't see any connection attempts at all suggests, that you have either entered the wrong hostname/IP in your applications or fail2ban blocked the connection.

For the latter, disable fail2ban, properly restart the container (`docker-compose up --force-recreate`) and see, if your applications are now able to send mails. If that works, it's likely, that fail2ban blocked some internal address (docker gateway), which your applications use to connect.

**Comment by ontheair81** [comment]
I already created a long answer with all the details, configurations and compose files. When I decided to double-check the blocked IPs by fail2ban again, as @casperklein recommended - what I have done already before.

And guess what... I did miss the fact that there was a blocked IP from the docker private range. Just unbanned it and everything works fine.

I am sorry for the noise... But a big "thank you" to @polarathene and @casperklein for the help and food for thoughts. You brought me back on the right track after dealing with this "problem" for hours.
Thanks!
